python_modules/dagster/dagster/_core/definitions/multi_dimensional_partitions.py

```diff
@@ -0,0 +1,52 @@
+from typing import List, Mapping, NamedTuple
+
+import dagster._check as check
+
+
+class PartitionDimensionKey(
+    NamedTuple("_PartitionDimensionKey", [("dimension_name", str), ("partition_key", str)])
+):
+    """
+    Representation of a single dimension of a multi-dimensional partition key.
+    """
+
+    def __new__(cls, dimension_name: str, partition_key: str):
+        return super(PartitionDimensionKey, cls).__new__(
+            cls,
+            dimension_name=check.str_param(dimension_name, "dimension_name"),
+            partition_key=check.str_param(partition_key, "partition_key"),
+        )
+
+
+class MultiPartitionKey(str):
+    """
+    A multi-dimensional partition key stores the partition key for each dimension.
+    Subclasses the string class to keep partition key type as a string.
+
+    Contains additional methods to access the partition key for each dimension.
+    Creates a string representation of the partition key for each dimension, separated by a pipe (|).
+    Orders the dimensions by name, to ensure consistent string representation.
+    """
+
+    dimension_keys: List[PartitionDimensionKey] = []
+
+    def __new__(cls, partition_dimension_mapping: Mapping[str, str]):
```

Comment:
A name like `keys_by_dimension` would be more clear and more consistent with the property